pkg/bpf/prog.go

```diff
@@ -67,3 +74,72 @@ func (t ProgType) String() string {
 
 	return "Unknown"
 }
+
+type attrProg struct {
+	startID   uint32
+	progID    uint32
+	nextID    uint32
+	openFlags uint32
+}
+
+// ProgInfo holds values from the upstram struct bpf_prog_info.
+// From: https://github.com/torvalds/linux/blob/master/include/uapi/linux/bpf.h#L1006
+type ProgInfo struct {
+	ProgType     ProgType
+	ID           uint32
+	MapIDs       uint64
+	CreatedByUID uint32
+	LoadTime     uint64 /* ns since boottime */
+	Name         string
+}
+
+type attrObjInfo struct {
+	bpfFD   uint32
+	infoLen uint32
+	info    uint64
+}
+
+// GetNextID takes a program ID and returns the next program ID.
+func GetNextID(start, next unsafe.Pointer) error {
+	attr := attrProg{
+		startID: uint32(uintptr(start)),
+		nextID:  uint32(uintptr(next)),
+	}
+
+	ret, _, err := unix.Syscall(unix.SYS_BPF, BPF_PROG_GET_NEXT_ID, uintptr(unsafe.Pointer(&attr)), unsafe.Sizeof(attr))
+	if ret != 0 || err != 0 {
+		return fmt.Errorf("Unable to get next id: %v", err)
+	}
+
+	return nil
+}
+
+// GetFDByID returns the file descriptor for the program id.
+func GetFDByID(id uint32) (int, error) {
+	attr := attrProg{
+		progID: uint32(uintptr(id)),
+	}
+
+	fd, _, err := unix.Syscall(unix.SYS_BPF, BPF_PROG_GET_FD_BY_ID, uintptr(unsafe.Pointer(&attr)), unsafe.Sizeof(attr))
+	if fd < 0 || err != 0 {
+		return int(fd), fmt.Errorf("Unable to get fd for program id %d: %v", id, err)
+	}
+
+	return int(fd), nil
+}
+
+// GetObjInfoByFD gets the bpf program info from it's file descriptor.
+func GetObjInfoByFD(fd int, info unsafe.Pointer, size uintptr) error {
```

Comment:
Can the `info` argument here be a `ProgInfo` pointer or something instead?

Reply:
done :)